### DISCUSSION TITLE
Can I use Service/Repo Pattern with inheritance?

### DISCUSSION THREAD

**Discussion by yogeshgalav** [laravel/framework#51589] [Ideas]
Php supports multilevel inheritance, so can I inherit Service/Repo classes with Models.
for e.g. 
```class UserService extends User(model) ```
```class PostRepo extends Post(model) ```

- It will offload the logic from controller and model as usual but resides in between.
- With this pattern I can use all the model properties from Service/Repo classes as well reducing multiple instance or class usage.
- Inside service/repo classes we can write db queries and use other service/repo classes inside each other as well.
- Models will remain clean consisting only of casting, relationships, scopes and modifiers.
- We can use only one service/repo class per model reducing multiple inheritance in service/repo class, hence reducing duplicate functions.

What's the problem with approach? Please Suggest.
If there's none should Laravel framework adapt this pattern by default?

**Comment by henzeb** [comment]
The only thing I can think of is the table name that is derived from the classname. The only way to find out is to try out.

But why not services or repository pattern? why is that stressful, and why are Helper classes not? Because in my book, Helper classes are just repository classes that happens to extend the model they are the repository of. And if you are going to create more than one Helper class per model, it's just a service that happens to extend the model they are a service for.

**Reply by yogeshgalav** [reply]
The thing i suppose which makes it better is inheritance through which I don't have to use different class or object in controller.
and I'm going to extend only one helper class per model for the same reason. So only multilevel inheritance, no multiple inheritance.

**Reply by ghost** [reply]
Some say there should be no queries in helpers:) But I don't see that as a live or die rule.
Also a helper is not instantiated and is usually used as static. If you do it like this you will transform the helper to an instance.

**Reply by yogeshgalav** [reply]
Question Modified.

**Reply by ghost** [reply]
@yogeshgalav you will have collisions between function names in service and in model.

EDIT. What is wrong with controller service model?
Look here how simple and short codded it can be:

https://github.com/macropay-solutions/laravel-crud-wizard-demo/blob/production/app/Models/Operation.php
https://github.com/macropay-solutions/laravel-crud-wizard-demo/blob/production/app/Services/OperationsService.php
https://github.com/macropay-solutions/laravel-crud-wizard-demo/blob/production/app/Http/Controllers/OperationsController.php

**Reply by yogeshgalav** [reply]
@marius-mcp Why did they used complex dependency injection when simply inheritance is available?
function name shouldn't conflict because models should only contain relationships, scope and getter,setters.

**Reply by henzeb** [reply]
With models you don't need to inject the model on your service/repo

**Reply by ghost** [reply]
@yogeshgalav the Eloquent model has update function for example. That will collide with service function:)

What do you mean by DI? In that example that I gave?
That is just a demo. You can put new service(); if \resolve(Service::class) bothers you.

UPDATE
also your model will be instantiated for each row from the db result set when you do something like 
`ModelServiceThatExtendsModel::query()->limit(10)->get();`

 Why put code there that will not be used?

**Reply by yogeshgalav** [reply]
@marius-mcp 
About update function override, Isn't that ideal
If User model has update function and UserRepo has update function too, which function would your fellow developer call?
It needs research, right? because it's confusing.

while in inheritance it will be override so no confusion. You can even call Model function from middle inherited class right?

**Reply by ghost** [reply]
You can go this road and you will see if you stumble on something or not. I would not do it. A model is a mode and a service is a service. Either way you need 2 classes. The argument that you need to have the model in the service is the only "gain" in your approach.

**Comment by yogeshgalav** [comment]
@marius-mcp 
In Laravel which code will run fast and efficient?
Getting class dependency by inheritance 
Or
Using resolve() for Manual Dependency Resolution?

In my php code, if
Dog class extends Mammal class which extends animal  class
will animal, mammal class always occupy memory even if I call function of animal class from object of dog?

**Reply by yogeshgalav** [reply]
chatgpt says
PHP uses a technique called “lazy loading” for methods. It means that the method code is loaded into memory only when the method is invoked.
When you call a method on an object (e.g., calling an Animal method from a Dog object), PHP looks up the method in the inheritance chain and loads it if necessary.

same case If I instantiate service or repo class in controller constructor.
